util/src/main/java/com/github/mylibrelab/text/DynamicText.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2020 MyLibreLab
+ * Based on MyOpenLab by Carmelo Salafia www.myopenlab.de
+ * Copyright (C) 2004  Carmelo Salafia cswi@gmx.de
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.github.mylibrelab.text;
+
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Implementation for {@link Text} which is based on a {@link Supplier<String>}.
+ */
+public class DynamicText implements Text {
+
+    private final Supplier<String> textSupplier;
+
+    /**
+     * Create a new {@link DynamicText} which determines it's text value based on the given supplier.
+     *
+     * @param textSupplier the text supplier.
+     */
+    public DynamicText(@NotNull final Supplier<String> textSupplier) {
+        this.textSupplier = textSupplier;
+    }
+
+    @NotNull
+    @Override
+    public String getText() {
+        return textSupplier.get();
+    }
+
+    @Override
+    public String toString() {
+        return getText();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        DynamicText that = (DynamicText) o;
+        return Objects.equals(textSupplier, that.textSupplier);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(textSupplier);
```

Comment:
> I noticed we are calling `Objects.hash()` for calculating hash codes, which is completely fine, it's just worth noting that it's performance can be improved, if deemed desirable.

`String` already cash their hashcode internally.

> For example, in case of `ConstantText` we can cache values and for `DynamicText` some sort of lazy initialization is possible.

What sort of lazy initialization do you have in mind? Most likely the `Supplier`s don't implement a custom `#hasCode` method and the identity hashcode is already cached by the jvm.

> But because both classes only have one significant field, it's best to use `Objects.hashCode()` anyway, that way we avoid unnecessary overhead of array creation.
> 
> Also, another thing worth noting about `hash()`: "Warning: When a single object reference is supplied, the returned value does not equal the hash code of that object reference. This value can be computed by calling `hashCode(Object).`" (hash of array containing object != object hash)

Thanks for pointing that out. I guess the generated `hashCode` implementation of IntelliJ doesn't optimize for single fields.

Reply:
Sorry for not being specific enough about the first part. I meant  to note that in general, if performance is important, we should optimize `Object.hash()`, because it can get costly pretty fast with all the array creations and possible boxing/unboxing. Just another thing that's maybe worth mentioning in dev docs. 

For single fields we should always use `hashCode()` because it avoids that cost completely.

Reply:
Oh yes definitely. I changed the implementation for `InternationalizedText` too. Somehow I assumes there would be overloads for some small amount of values like `List#of` to avoid array creation.

Reply:
Yup, `hash()` is not really smart, it's just an easy utility method for multiple fields:

`Generates a hash code for a sequence of input values. The hash code is generated as if all the input values were placed into an array, and that array were hashed by calling Arrays.hashCode(Object[]). `

So yeah, the problem is array creation, and no caching optimizations for things you mentioned, like strings. Everything is tossed into the array, and the whole array is hashed. Which is completely fine for some use cases, but we should be aware of that cost because it can be significant.